### PR TITLE
fix: duplicate user email

### PR DESCRIPTION
**Pull request by wansir**
Signed-off-by: hongming <talonwan@yunify.com>

### CODE REVIEW

File: pkg/apiserver/iam/im.go [hunk 1]
```diff
@@ -105,7 +105,7 @@ func UpdateUser(req *restful.Request, resp *restful.Response) {
 	err := req.ReadEntity(&user)
 
 	if err != nil {
-		resp.WriteHeaderAndEntity(http.StatusBadRequest, errors.Wrap(err))
+		resp.WriteHeaderAndEntity(http.StatusConflict, errors.Wrap(err))
```
Comment (runzexia) [inline pkg/apiserver/iam/im.go]: should be bad request?

File: pkg/apiserver/iam/im.go [hunk 2]
```diff
@@ -143,6 +143,11 @@ func UpdateUser(req *restful.Request, resp *restful.Response) {
 	result, err := iam.UpdateUser(&user)
 
 	if err != nil {
+		if ldap.IsErrorWithCode(err, ldap.LDAPResultEntryAlreadyExists) {
+			resp.WriteHeaderAndEntity(http.StatusBadRequest, errors.Wrap(err))
+			return
+		}
+
 		resp.WriteHeaderAndEntity(http.StatusInternalServerError, errors.Wrap(err))
```
Comment (runzexia) [inline pkg/apiserver/iam/im.go]: maybe log here

File: pkg/models/iam/im.go [hunk 3]
```diff
@@ -883,6 +884,28 @@ func UpdateUser(user *models.User) (*models.User, error) {
...
+			[]string{"uid", "mail"},
+			nil,
+		)
+		result, err := conn.Search(userSearchRequest)
+		if err != nil {
+			glog.Error(err)
+			return nil, err
+		}
+		if len(result.Entries) > 1 {
+			err = ldap.NewError(ldap.ErrorDebugging, fmt.Errorf("email is duplicate"))
```
Comment (zryfish) [inline pkg/models/iam/im.go]: I think we'd better have email in error message, something like
`fmt.Errorf(duplicated email: %s, email)`

File: pkg/models/iam/im.go [hunk 4]
```diff
@@ -883,6 +884,28 @@ func UpdateUser(user *models.User) (*models.User, error) {
...
+			glog.Error(err)
+			return nil, err
+		}
+		if len(result.Entries) == 1 && result.Entries[0].GetAttributeValue("uid") != user.Username {
+			err = ldap.NewError(ldap.LDAPResultEntryAlreadyExists, fmt.Errorf("email is duplicate"))
```
Comment (zryfish) [inline pkg/models/iam/im.go]: same here

File: pkg/models/iam/im.go [hunk 5]
```diff
@@ -883,6 +884,28 @@ func UpdateUser(user *models.User) (*models.User, error) {
...
+			[]string{"uid", "mail"},
+			nil,
+		)
+		result, err := conn.Search(userSearchRequest)
+		if err != nil {
+			glog.Error(err)
+			return nil, err
+		}
+		if len(result.Entries) > 1 {
+			err = ldap.NewError(ldap.ErrorDebugging, fmt.Errorf("email is duplicate: %s", user.Email))
```
Comment (zryfish) [inline pkg/models/iam/im.go]: `duplicated`

File: pkg/models/iam/im.go [hunk 6]
```diff
@@ -883,6 +884,28 @@ func UpdateUser(user *models.User) (*models.User, error) {
...
+			glog.Error(err)
+			return nil, err
+		}
+		if len(result.Entries) == 1 && result.Entries[0].GetAttributeValue("uid") != user.Username {
+			err = ldap.NewError(ldap.LDAPResultEntryAlreadyExists, fmt.Errorf("email is duplicate: %s", user.Email))
```
Comment (zryfish) [inline pkg/models/iam/im.go]: duplicated

**Pull request merged**